cmd/containerd-shim-runc-v2/task/service.go

```diff
@@ -707,18 +730,94 @@ func (s *service) send(evt interface{}) {
 	s.events <- evt
 }
 
-// s.mu must be locked when calling handleProcessExit
-func (s *service) handleProcessExit(e runcC.Exit, c *runc.Container, p process.Process) {
-	if ip, ok := p.(*process.Init); ok {
-		// Ensure all children are killed
-		if runc.ShouldKillAllOnExit(s.context, c.Bundle) {
-			if err := ip.KillAll(s.context); err != nil {
-				log.G(s.context).WithError(err).WithField("id", ip.ID()).
-					Error("failed to kill init's children")
+// handleInitExit processes container init process exits.
+// This is handled separately from non-container init exits, because there
+// are some extra invariants we want to ensure in this case, namely:
+// - the init process exit MUST be the last exit we publish for a given container
+// - IF a container init process is reaped, we MUST ALWAYS publish it's exit
+// This is achieved by:
+// - waiting for all pending execs to either successfully start or early-exit
+// (this happens before handleInitExit is called)
+// - killing all running container processes
+// - waiting for the running process count (for a container) to go to 0, or
+// timing out.
+// - removing all processes from s.running so that we don't publish late
+// exec exits, and publishing the init exit.
+func (s *service) handleInitExit(e runcC.Exit, c *runc.Container, p *process.Init) {
+	s.lifecycleMu.Lock()
+	defer s.lifecycleMu.Unlock()
+
+	// kill all running container processes
+	if err := p.KillAll(s.context); err != nil {
```

Comment:
I'm not thrilled about execing a subprocess while holding lifecycleMu. Does the kill-all need to be synchronized? Exec starts for the container should already be blocked from starting at this point so I don't see the need to hold the lock while doing it.

Reply:
Good point! I don't see why this should need to be synchronized.